Can't build testscanner.exe on Windows

I got this errors

```
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x6428): undefined reference to `JSonScanner::JSonScanner(QIODevice*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x64a7): undefined reference to `JSonScanner::yylex(QVariant*, yy::location*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x6536): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x6565): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x6733): undefined reference to `JSonScanner::JSonScanner(QIODevice*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x674f): undefined reference to `JSonScanner::allowSpecialNumbers(bool)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x67ce): undefined reference to `JSonScanner::yylex(QVariant*, yy::location*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x67f2): undefined reference to `JSonScanner::yylex(QVariant*, yy::location*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x6e76): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x7149): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x73e3): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x7661): undefined reference to `JSonScanner::JSonScanner(QIODevice*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x7676): undefined reference to `JSonScanner::allowSpecialNumbers(bool)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x76f5): undefined reference to `JSonScanner::yylex(QVariant*, yy::location*)'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x77d0): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x7b53): undefined reference to `JSonScanner::~JSonScanner()'
CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj):testscanner.cpp:(.text+0x7b84): undefined reference to `JSonScanner::~JSonScanner()'
C:/Qt/Qt5.3.1/Tools/mingw482_32/bin/../lib/gcc/i686-w64-mingw32/4.8.2/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\testscanner.dir/objects.a(testscanner.cpp.obj): bad reloc address 0x4 in section `.data'
```

Problem in JsonScanner class. It's internal class which hasn't QJSON_EXPORT. Windows requires some special attributes for classes/functions in shared libs. So JsonScanner must be defined with QJSON_EXPORT or testscanner must be linked against ./src/CMakeFiles/qjson.dir/json_scanner.cpp.obj.

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -23,6 +23,13 @@ set (qjson_HEADERS parser.h parserrunnable.h qobjecthelper.h serializer.h serial
 # Required to use the intree copy of FlexLexer.h
 INCLUDE_DIRECTORIES(.)
 
+# Special hack to fix scanner test compilation on Windows
+# JsonScanner class is not for export. So need to make static
+# lib what can be compiled against scanner test.
+if(WIN32 AND QJSON_BUILD_TESTS AND BUILD_SHARED_LIBS)
+  add_library(qjson_scanner STATIC json_scanner.cpp)
+endif()
+
 add_library (qjson ${qjson_SRCS} ${qjson_MOC_SRCS} ${qjson_HEADERS})
 IF (Qt5Core_FOUND)
   target_link_libraries( qjson ${Qt5Core_LIBRARIES})

--- a/tests/scanner/CMakeLists.txt
+++ b/tests/scanner/CMakeLists.txt
@@ -31,12 +31,17 @@ FOREACH(test ${UNIT_TESTS})
     ${test}.cpp
   )
 
+  if(WIN32 AND BUILD_SHARED_LIBS)
+    set(QJSON_SCANNER qjson_scanner)
+  endif()
+
   ADD_FILE_DEPENDENCIES(${test}.cpp ${MOC_SOURCE})
   TARGET_LINK_LIBRARIES(
     ${test}
     ${QT_LIBRARIES}
     ${TEST_LIBRARIES}
     qjson
+    ${QJSON_SCANNER}
   )
   if (QJSON_TEST_OUTPUT STREQUAL "xml")
     # produce XML output

